Evaluate independent instructions only once

AYadrov opened the pull request:
This PR adds an optimization that allow us to evaluate some operations only once per the whole expression.

The key idea is: when operations do not depend on the arguments - they can be evaluated only once at compile stage.
Moreover, they can be evaluated at the maximum precision, so that the reevaluation would never be required.

Therefore, their outputs will be implemented as a part of `hint` - referring to already precise-enough results.

~Hmm, what if user will change precision? then, we won't reevaluate these operations anyways. that's bad, probably the logic should be changed before merging so that the reevaluation at higher precision would be possible~

AYadrov commented on eval/run.rkt:
```diff
@@ -46,18 +47,33 @@
         [#t
-         (parameterize ([bf-precision precision])
-           (apply-instruction instr vregs))]
+         (define start (current-inexact-milliseconds))
```
Track execution only for operations that are actually executed

AYadrov commented on infra/run-baseline.rkt:
```diff
@@ -156,10 +156,11 @@
+  (define start-prec (+ (discretization-target (last discs)) (*base-tuning-precision*)))
   (define precisions
-    (make-vector (vector-length instructions))) ; vector that stores working precisions
+    (make-vector (vector-length instructions) start-prec)) ; vector that stores working precisions
```
We need some starting precisions for `default-hint` that would make sure to evaluate independent instructions once

AYadrov commented on eval/run.rkt:
```diff
@@ -46,18 +47,33 @@
+         (rival-machine-record machine name n precision time)
+         res]
+        [(box old-precision)
+         (match (> precision old-precision)
```
If assigned precision is higher than precision with what the independent instruction has been evaluated already - reevaluate

pavpanchekha commented:
I've been thinking that we should fix the maximum precision during the `compile` step; then we wouldn't have this issue with the maximum precision changing. What do you think of that? It would also be helpful for `add-bang` since it would mean we can pre-allocate a register of maximum size.

AYadrov commented:
Oh, yeah, I fixed this thing this morning. 
I implemented it this way so that we start evaluation of _independent_ instructions with a low precision and we can reevaluate if backward pass assigns ever a precision that is higher.
I keep track of it using hints, hint for _independent_ operation is a `box` with precision. 
Evaluation loops checks whether the current precision is higher than `box` - if so, reevaluate, update the `box`.
Otherwise, just read from `registers` vector

AYadrov commented:
Quite non-obvious bug has been observed.
What happens is:
1) we calculate independent instructions at precision `t`
2) these independent instructions end up requiring more precision than `t`
3) some instructions get more bits, some instructions stay the same.

The thing is, once some instructions get reexecuted at higher precision - it means that all the parent should be reexecuted as well. It is clear why children received more precision - error exists. Therefore, even despite that we have evaluated some independent instructions at precision `t`, if their children (they are also independent btw) get reexecuted - parent also should be. 

I made a quick fix for that, I do not distinct, which parent, which children. Just once I see that an independent instruction needs a reexecution - reexecuted all following instructions as well - some of them can be a parent of that reexecuted one. 
It should work. Slightly greedy - but this case is gonna be rare

pavpanchekha commented:
I think this bug is because the independent subexpr thing should actually use the repeats mechanism, not the hints mechanism, for constant sub expressions.

pavpanchekha commented:
Here’s my proposal: merge all this stuff into an “oopsla” branch without waiting for review/cleanup. I’ll make that the default on nightly. Then we’ll do clean up after the deadline.
